### ISSUE TITLE
Const generics: compile error if array length is u8

### ISSUE THREAD

**Issue by scvalex** [rust-lang/rust#104949]
I tried this code:

```rust
pub struct MaxLengthString<const N: u8> {
    len: u8,
    buf: [u8; N],
}
```

And I got this compilation error:

```
error[E0308]: mismatched types
 --> src/lib.rs:3:15
  |
3 |     buf: [u8; N],
  |               ^ expected `usize`, found `u8`

For more information about this error, try `rustc --explain E0308`.
```

This works fine if the const parameter is of type `usize`, but I think it should work for the other unsigned types that are smaller than `usize`.  Or at the very least, maybe there should be a different error message that says this is definitely not supported.

### Meta

`rustc --version --verbose`:
```
rustc 1.67.0-nightly (8681d4cff 2022-11-25)
binary: rustc
commit-hash: 8681d4cffcd23bbe619984ab62772a91827a40dc
commit-date: 2022-11-25
host: x86_64-unknown-linux-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```

**Comment by Nilstrieb** [comment]
Only allowing `usize` is intended, but improvements to error messages are of course always welcome :)

**Comment by jruderman** [comment]
This alternative works, but I don't know if it's appropriate for diagnostics to suggest it:

```rust
#![feature(generic_const_exprs)]
#![allow(incomplete_features)]

pub struct MaxLengthString<const N: u8> 
where [(); N as usize]:
{
    len: u8,
    buf: [u8; N as usize],
}
```

**Comment by estebank** [comment]
[The logic to produce that suggestion exists in rustc, but it is dormant under the assumption that we want to move people as much as possible towards `into` and `try_into`](https://github.com/rust-lang/rust/blob/4439f1f6a6ac2e7b7a303717855addf505a9e591/compiler/rustc_hir_typeck/src/demand.rs#L1072-L1080), but [we already know they are not available in const context](https://github.com/rust-lang/rust/blob/4439f1f6a6ac2e7b7a303717855addf505a9e591/compiler/rustc_hir_typeck/src/demand.rs#L1184), so we might want to start giving them in that case.

**Comment by jruderman** [comment]
Why aren't `into` and `try_into` available in const?

**Comment by Nilstrieb** [comment]
Const traits are far from stable.

**Comment by gyscos** [comment]
So to confirm, currently, there is no way on stable to turn a `const N: u8` into an array size?

**Comment by estebank** [comment]
@gyscos correct, [you can only do it on nightly](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=42e88cbeedc43a04f4fe6e70b772e4e3).

Triage: no change.
